File: src/engine/rng.ts
// src/engine/rng.ts — pure, deterministic seeded randomness used INSIDE reducers.
// The impure entropy that produces seeds lives in services/random.ts. No Math.random / Date here.

/** mulberry32: returns a generator producing floats in [0,1) deterministically from `seed`. */
export function rng(seed: number): () => number {
  let a = seed >>> 0;
  return function next(): number {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Fisher–Yates using a seed; returns a NEW shuffled array (input untouched). */
export function shuffle<T>(items: readonly T[], seed: number): T[] {
  const arr = items.slice();
  const next = rng(seed);
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(next() * (i + 1));
    const tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
  }
  return arr;
}

/** Deterministic integer in [min, max] inclusive. */
export function int(min: number, max: number, seed: number): number {
  let lo = Math.ceil(min);
  let hi = Math.floor(max);
  if (hi < lo) [lo, hi] = [hi, lo];
  const next = rng(seed);
  return lo + Math.floor(next() * (hi - lo + 1));
}

/** Deterministic pick of one item. Throws on empty input. */
export function pick<T>(items: readonly T[], seed: number): T {
  if (items.length === 0) throw new Error('pick: empty array');
  return items[int(0, items.length - 1, seed)];
}

/** Derive a child seed deterministically (chain multiple draws from one action seed). */
export function deriveSeed(seed: number, salt: number): number {
  let h = (seed >>> 0) ^ Math.imul(salt >>> 0, 0x9e3779b1);
  h = Math.imul(h ^ (h >>> 16), 0x45d9f3b);
  h = Math.imul(h ^ (h >>> 16), 0x45d9f3b);
  return (h ^ (h >>> 16)) >>> 0;
}

File: android-native/app/src/main/java/com/gamenight/party/engine/Rng.kt
package com.gamenight.party.engine

/**
 * Seeded, deterministic randomness — a Kotlin port of src/engine/rng.ts (mulberry32).
 * Games stay pure by threading seeds through actions, exactly like the web engine: get a fresh
 * seed from the host (see [freshSeed]) at the impure boundary, then everything downstream is
 * reproducible.
 */
class Rng(seed: Int) {
    private var state: Int = seed

    /** Next float in [0, 1). Bit-for-bit equivalent to the web's mulberry32. */
    fun next(): Double {
        state += 0x6D2B79F5
        var t = (state xor (state ushr 15)) * (1 or state)
        t = (t + ((t xor (t ushr 7)) * (61 or t))) xor t
        return (t xor (t ushr 14)).toUInt().toDouble() / 4294967296.0
    }

    /** Inclusive integer in [min, max]. */
    fun int(min: Int, max: Int): Int = min + (next() * (max - min + 1)).toInt()

    /** A uniformly random element (throws on empty, like the web `pick`). */
    fun <T> pick(items: List<T>): T {
        require(items.isNotEmpty()) { "pick() on empty list" }
        return items[(next() * items.size).toInt()]
    }

    /** A new shuffled copy (Fisher–Yates). */
    fun <T> shuffle(items: List<T>): List<T> {
        val out = items.toMutableList()
        for (i in out.indices.reversed()) {
            val j = (next() * (i + 1)).toInt()
            val tmp = out[i]; out[i] = out[j]; out[j] = tmp
        }
        return out
    }
}

/** Derive a stable child seed from a parent seed + salt (mirrors rng.ts deriveSeed). */
fun deriveSeed(seed: Int, salt: Int): Int {
    var h = seed xor (salt * -0x61c88647) // golden-ratio mix
    h = (h xor (h ushr 16)) * -0x7ee3623b
    h = (h xor (h ushr 13)) * -0x3d4d51cb
    return h xor (h ushr 16)
}
